provider.go

```diff
@@ -0,0 +1,194 @@
+package kindsys
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+
+	"cuelang.org/go/cue"
+	"github.com/grafana/thema"
+)
+
+const providerPackageName = "provider"
```

Comment:
Defining a provider currently requires it to be defined in a package named `provider`. Not sure if this is satisfactory/what we want?

Reply:
On the one hand, yes, it's typically important for standardized loaders that we pick one package name and require that it be the CUE package in which folks write certain things.

On the other, if we're going to have a few different "implementations" of provider (e.g. plugins vs. apps) it seems likely those implementations may want to make different choices about what that expected package name is.

Reply:
> It can't, it has to know _a priori_ what it's looking for. Which means it's part of the user-facing requirement: if the plugins loader decides to expect `package grafanaplugin`, that has to be publicly documented, enforced by tooling, etc. That's what [we currently do in grafana/grafana's pfs](https://github.com/grafana/grafana/blob/09895c26b6bd2c4e87ea780ff8bc16dc2f8c407e/pkg/plugins/pfs/pfs.go#L156). If app sdk is expecting something else, it has to document that where appropriate, separately.
> 
> So, this is a _very_ low-level bit of architecture. It may be useful for us to not hardcode it at the level of kind provider, but only because we're planning on making that decision in the very next layer up. There's no end-user choice involved here - the package name to use must be firmly decided long before it makes its way to any app/plugin author.

Okay. But isn't one of the goals with the kind provider that we can write tooling (codegen) targeting kind provider rather than targeting the very next layer up for us to be able to share common tooling (codegen etc)? If not, the tooling (codegen) needs to know what package name to use based on what the provider is which seems not great from the perspective of schema registry where you can have a mix of schemas originating from different providers etc. I might have understood this completely wrong? :) 

> That's almost certainly the approach to take for now while we're transitioning. The main nasty bit here is yet another naming thing - what name do we choose if we inject the `cue.mod`? The answer i [fell back on in pfs](https://github.com/grafana/grafana/blob/09895c26b6bd2c4e87ea780ff8bc16dc2f8c407e/pkg/plugins/pfs/pfs.go#L241) - `grafana.com/plugins/$PLUGIN_ID` is...ok, but also implies how this name should probably be tied to package identity.

Yes, I've noticed we do that in pfs. Sure, it could make sense. This is still dynamically happen at runtime/load time. But I guess one thing I haven't understood is whether or how this generated cue.mod file is becoming a part of a schema registry in that case? 

> That led me to Option 1.

Based on my above comments I wrote sounds like enforcing/hardcode the provider package name at the level of kind provider is the way forward and was what I originally did, e.g.
```
const providerPackageName = "provider"
```

This feels like the easiest way forward for now at least.

Reply:
> But I guess one thing I haven't understood is whether or how this generated cue.mod file is becoming a part of a schema registry in that case?

This one i can answer definitively - it should not. If we reuse the same loader funcs that dynamically inject a cue.mod ~everywhere, there's no reason to ever persist the results of that dynamic injection.

> Okay. But isn't one of the goals with the kind provider that we can write tooling (codegen) targeting kind provider rather than targeting the very next layer up for us to be able to share common tooling (codegen etc)? If not, the tooling (codegen) needs to know what package name to use based on what the provider is which seems not great from the perspective of schema registry where you can have a mix of schemas originating from different providers etc. I might have understood this completely wrong? :)

Mmm, interesting point...i don't think you've misunderstood, i think i just missed this :) It's certainly relevant to the `LoadProvider` func in this PR. 

I do still think it's preferable to not enforce the package name at quite this level. Unfortunately, unpacking what "this level" means gets us into the weeds with CUE loading.

I'm not sure how easily i can suggest a path here, given the number of considerations - it may be that it's just easiest for me to hack on this a bit to arrange the pieces sanely. But i'll start at least by outlining some names and patterns, and we can see what makes sense as the next step.

CUE itself establishes that there are three representational states that the bytes representing a bunch of user-provided CUE expressions can be in. Let's refer to these as (my names):

* Raw - the literal raw bytes from a user. Might not even be syntactically valid CUE. 
* Loaded - the bytes have been parsed by `[load.Instances](https://pkg.go.dev/cuelang.org/go@v0.5.0/cue/load#Instances)`, and are now represented as a [`build.Instance`](https://pkg.go.dev/cuelang.org/go@v0.5.0/cue/build#Instance).
* Built - a `build.Instance` has been loaded into the CUE graph by a call like [`cue.Context.BuildInstance`](https://pkg.go.dev/cuelang.org/go/cue@v0.5.0#Context.BuildInstance), resulting in a `cue.Value`. The `cue.Value.Err()` method on this result indicates if there was a semantic problem in the original inputs.

Or, really handwaving, we could express in a flow diagram that there are some common types, with common verbs that transition between the types:

```
[]byte -- "Load()" --> *build.Instance -- "Build()" --> cue.Value
```

Let's try to think of each of these (loosely, they are not perfect fits) in the spirit of "[Parse, Don't Validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)". With that in mind, let's note that even if we have a semantically correct `cue.Value` from CUE's perspective, it's not necessarily a valid instance of the thing _we_ want - in this case a Provider. So we need a fourth step, for which we've generally been using the word "Bind":

```
[]byte -- "Load()" --> *build.Instance -- "Build()" --> cue.Value -- "Bind()" --> (our ecosystem type)
```

[BindLineage](https://pkg.go.dev/github.com/grafana/thema#BindLineage) solidly falls in this category. [BindCore](https://pkg.go.dev/github.com/grafana/kindsys#BindCore) and its siblings sorta do (and i'd like to take them from sorta -> solid).

For now at least, let's call anything that fits into this general multi-step process, from bytes to some native Go type, a "kindsys compiler."

It's expected that we have more than one kindsys compiler. It's also expected that we should be able to make those compilers out of reusable parts, if we create those parts carefully. [`load.InstanceWithThema`](https://pkg.go.dev/github.com/grafana/thema/load@v0.0.0-20230612151146-2a2af6b671a5#InstanceWithThema) is the proof there - it wraps CUE's `load.Instances`, but the thema wrapper is all we ever need in kindsys, or grafana, or app sdk.

I don't have a totally fixed idea yet of what all these passes should be. But i'm decently sure that trying to follow a general rule of keeping these passes as logically separate functions has the best chance of letting us iterate towards something good.

i don't imagine this is clear enough to immediately unblock progress - apologies! - we can plan to circle back on it.